### ISSUE TITLE
The ipa size grows too fast due to Flutter's incompact machine code instructions

### ISSUE THREAD

**Issue by limengyun2008** [flutter/flutter#40345]
## Steps to Reproduce

We've noticed that the Flutter compilation production size grows too fast.

one evidence is after adding 1300+ lines of dart code, the ipa size increased by 200kb+, we are very sure that OC's products will never be that big.

so we made a test:

this is a simple OC function:
```c
// OC
- (void) ocTestInt {
    int a = 3 + 4;
    NSLog(@"%d", a);
}
```
 assembly
![image](https://user-images.githubusercontent.com/2217633/64773665-5a11e680-d585-11e9-808f-f62437668369.png)

then write the same function, but using dart:
```dart
// Flutter
void flutterTestInt() {
  int a = 3 + 4;
  print(a);
}
```

assembly
![image](https://user-images.githubusercontent.com/2217633/64773919-cab90300-d585-11e9-9361-724c37854c11.png)

.quad is not the real machine code of the function so we need disassemble it
![image](https://user-images.githubusercontent.com/2217633/64774016-fb00a180-d585-11e9-8d45-2ac6d6958655.png)

below is the real arm64 instructions
![image](https://user-images.githubusercontent.com/2217633/64774118-28e5e600-d586-11e9-840f-a60a9e24d94b.png)

32 VS 11, dart's instructions' number is three times bigger than OC
We've tested other functions, and the result is the same.

I think this is the final reason that why our .ipa grows too fast. could some guy explain it for me?

**Comment by limengyun2008** [comment]
@zoeyfan  I think that thread is talking about modular compiling, or engine size?
but what I am talking about is **our dart code's package size** : it grows too much faster than expected.
not really the same thing.

**Comment by xster** [comment]
Right, it's an orthogonal issue. I think the solution for this one is to better document gen_snapshot's internals. I also vaguely remember @mraleph having some debug outputs to correlate Dart code input size to AOT output size etc.

**Comment by xster** [comment]
@limengyun2008 can we repurpose this bug to a general Dart output diagnostics tool ask?

cc @a-siva

**Comment by a-siva** [comment]
I don't think we should re-purpose this bug for the generic output diagnostics tool. This specific to the question about why we see more instructions in the Dart case.

**Comment by a-siva** [comment]
@limengyun2008 thanks for your detailed report regarding this issue. Couple of questions
- Is the sample function you have above representative of most functions in your app (i.e you have a number of small functions).
- we expect the compiler to inline small functions into the caller and I was wondering if you are observing small functions not getting inlined.

Is it possible to add a couple of more examples of functions from your app to this issue.

**Comment by limengyun2008** [comment]
@a-siva 

actually, we made another test before.

## test 1(Copy & Paste)

we are being strong challenged by other team's leader who is responsible for package size now.

what he said is: I can accept 10M or 20M or 30M package size cost for once to try flutter, but I can not accept the tendency that all the next features will be implemented by dart. 200KB in OC means a very big feature, but I only see a simple single new page in the commit. if we don't stop this fucking flutter, package size will be uncontrollable. **we are not talking about package size, we are talking about package size increasement!**

so we made a test to verify his judgement about the size increasement between dart and oc. unfortunately, he is right. the result is: **TRIPPLE**.

below is a dart code sinppet
```dart
Widget func1() {
  return Text('blablabla');
}

Widget func2() {
  return Text('blablabla');
}

Widget func3() {
  return Column(children: <Widget>[
    Text('blablabla'),
    Text('blablabla'),
    Text('blablabla'),
    Icon(Icons.add),
  ],);
}
```
then continue doing one thing: copy, rename and invoke the functions. the size do not increase linearly, it will suddenly grow 16KB. finally we have a conclusion: 

16 KB ≈ 10 snippets ≈ 30 functions

what about OC? adding 30 functions like below = 5KB. 
```c
- (UIView *)func1 {
    return [[UIView alloc] initWithFrame:CGRectMake(0, 0, 0, 0)];
}
- (UIView *)func2 {
    return [[UIView alloc] initWithFrame:CGRectMake(0, 0, 0, 0)];
}
- (UIView *)func3 {
    [UIView animateWithDuration:1.0 animations:^{
        self.view.frame = CGRectMake(0, 0, 0, 0);
        self.view.backgroundColor = [UIColor redColor];
    }];
    return [[UIView alloc] initWithFrame:CGRectMake(0, 0, 0, 0)];
}
```

**16KB/5KB = 3**
oh, shit!

## test 2 (compre assembly)
test 1 is quite fair, but I still I can't believe my eyes. 

it occurs to us that we can compare the assembly to find out the reason, as I've mentioned before, it turn out that dart generates **TRIPLE** instructions compared with the same logic written in OC. 

It is the reason!

## back to your questions

> Is the sample function you have above representative of most functions in your app (i.e you have a number of small functions).

testInt is definitely not the real situation and our app's dart codes are quite normal, or at least, I think it is. 

> we expect the compiler to inline small functions into the caller and I was wondering if you are observing small functions not getting inlined.

inlining have nothing to do with this issue. actually i think inlining will enlarge the packages so I closed inlining(inline.cc, right?) before I do my tests.

> Is it possible to add a couple of more examples of functions from your app to this issue.

test 1 may be another example... and rewriting(from dart to oc) a real function of our app is impossible since it uses too many other dart libs/classes. it's difficult to make a compare test in real world and I don't think it's necessary. Because It's not an individual issue. it's a common issue that every function written in dart has the same problem: They are fat.

**Comment by mraleph** [comment]
@limengyun2008 

The actual core of the `flutterTestInt` function in fact has roughly the same size in Dart as it has in Objective-C.

32 bytes (8 "instructions") at the start of the function are instruction header (which Objective-C does not have) - a fixed price you pay per function, which can actually be eliminated. I have filed https://github.com/dart-lang/sdk/issues/38451 to track this work.

24 bytes (6 "instructions") at the end of the function are alignment padding. We align all instructions by 32-bytes for (legacy) performance reasons. Objective-C compiler does not seem to do that. I have filed https://github.com/dart-lang/sdk/issues/38452 to track this work.

The rest of the code in the function falls apart like this:

```
0x7fea2f6e8340    d14005f0               subi ip0, r15, 0x1000
0x7fea2f6e8344    927cee1f               andi csp, ip0, 0xfffffffffffffff0
0x7fea2f6e8348    a9bf79fd               stp fp, lr, [r15, #-16 ]!
0x7fea2f6e834c    aa0f03fd               mov fp, r15
        ;; CheckStackOverflow:8(stack=0, loop=0)
0x7fea2f6e8350    f9402750               ldrx ip0, [thr, #72]
0x7fea2f6e8354    eb1001ff               cmp r15, ip0
0x7fea2f6e8358    54000149               bls 0x7fea2f6e8380
        ;; PushArgument(v6)
0x7fea2f6e835c    d28001d0               movz ip0, 0xe
0x7fea2f6e8360    f81f8df0               strx ip0, [r15, #-8] !
        ;; StaticCall:14( print<0> v6)
0x7fea2f6e8364    94000000               bl 0x7fea2f6e8364
0x7fea2f6e8368    910021ef               addi r15, r15, 0x8
        ;; ParallelMove r0 <- C
0x7fea2f6e836c    f9405f40               ldrx r0, [thr, #184]   null
        ;; Return:18(v0)
0x7fea2f6e8370    aa1d03ef               mov r15, fp
0x7fea2f6e8374    a8c179fd               ldp fp, lr, [r15], #16 !
0x7fea2f6e8378    d65f03c0               ret lr
0x7fea2f6e837c    d4200000               brk 0x0
        ;; CheckStackOverflowSlowPath
0x7fea2f6e8380    94000000               bl 0x7fea2f6e8380
0x7fea2f6e8384    54fffece               b 0x7fea2f6e835c
```

This is 18 instructions vs 11 instructions of Objective-C compiler. 

Out of this 18 instructions 5 are dedicated to stack-overflow check, which serves both as a stack overflow detection mechanism and an interruption check, for GC and OOB message passing. Something that Objective-C does not have. 

This leaves us with 13 instructions - roughly the same size as Objective-C code. 

I do however realise that this core (13 instructions - 52 bytes) are wrapped in a lot of overhead (at least 56 bytes) and this is not counting some other metadata that Dart compiler associates with generated code - so indeed if you code consists of a lot of small functions that will bloat the binary rather fast compared to Objective-C binary. 

Some of this overhead is easy to remove, but some of it is not so easy. 

Let me follow up with analysis of your other reproduction.

**Comment by limengyun2008** [comment]
@mraleph 
Great! Very Clear!
looking forward to the future dart release.

And....Will your dartVM book(https://mrale.ph/dartvm/) be updated continuously? 
or, you are meant to update it yearly... lol..

**Comment by yuanhuihui** [comment]
> @limengyun2008
> 
> The actual core of the `flutterTestInt` function in fact has roughly the same size in Dart as it has in Objective-C.
> 
> 32 bytes (8 "instructions") at the start of the function are instruction header (which Objective-C does not have) - a fixed price you pay per function, which can actually be eliminated. I have filed [dart-lang/sdk#38451](https://github.com/dart-lang/sdk/issues/38451) to track this work.
> 
> 24 bytes (6 "instructions") at the end of the function are alignment padding. We align all instructions by 32-bytes for (legacy) performance reasons. Objective-C compiler does not seem to do that. I have filed [dart-lang/sdk#38452](https://github.com/dart-lang/sdk/issues/38452) to track this work.
> 
> The rest of the code in the function falls apart like this:
> 
> ```
> 0x7fea2f6e8340    d14005f0               subi ip0, r15, 0x1000
> 0x7fea2f6e8344    927cee1f               andi csp, ip0, 0xfffffffffffffff0
> 0x7fea2f6e8348    a9bf79fd               stp fp, lr, [r15, #-16 ]!
> 0x7fea2f6e834c    aa0f03fd               mov fp, r15
>         ;; CheckStackOverflow:8(stack=0, loop=0)
> 0x7fea2f6e8350    f9402750               ldrx ip0, [thr, #72]
> 0x7fea2f6e8354    eb1001ff               cmp r15, ip0
> 0x7fea2f6e8358    54000149               bls 0x7fea2f6e8380
>         ;; PushArgument(v6)
> 0x7fea2f6e835c    d28001d0               movz ip0, 0xe
> 0x7fea2f6e8360    f81f8df0               strx ip0, [r15, #-8] !
>         ;; StaticCall:14( print<0> v6)
> 0x7fea2f6e8364    94000000               bl 0x7fea2f6e8364
> 0x7fea2f6e8368    910021ef               addi r15, r15, 0x8
>         ;; ParallelMove r0 <- C
> 0x7fea2f6e836c    f9405f40               ldrx r0, [thr, #184]   null
>         ;; Return:18(v0)
> 0x7fea2f6e8370    aa1d03ef               mov r15, fp
> 0x7fea2f6e8374    a8c179fd               ldp fp, lr, [r15], #16 !
> 0x7fea2f6e8378    d65f03c0               ret lr
> 0x7fea2f6e837c    d4200000               brk 0x0
>         ;; CheckStackOverflowSlowPath
> 0x7fea2f6e8380    94000000               bl 0x7fea2f6e8380
> 0x7fea2f6e8384    54fffece               b 0x7fea2f6e835c
> ```
> 
> This is 18 instructions vs 11 instructions of Objective-C compiler.
> 
> Out of this 18 instructions 5 are dedicated to stack-overflow check, which serves both as a stack overflow detection mechanism and an interruption check, for GC and OOB message passing. Something that Objective-C does not have.
> 
> This leaves us with 13 instructions - roughly the same size as Objective-C code.
> 
> I do however realise that this core (13 instructions - 52 bytes) are wrapped in a lot of overhead (at least 56 bytes) and this is not counting some other metadata that Dart compiler associates with generated code - so indeed if you code consists of a lot of small functions that will bloat the binary rather fast compared to Objective-C binary.
> 
> Some of this overhead is easy to remove, but some of it is not so easy.
> 
> Let me follow up with analysis of your other reproduction.

wonderful, looking forward to your results

**Comment by mraleph** [comment]
I looked at `func1` (just instruction size). In Dart we get 36 instructions: 

```
        ;; B0
        ;; B1
        ;; Enter frame
0x10e859340    d14005f0               subi ip0, r15, 0x1000
0x10e859344    927cee1f               andi csp, ip0, 0xfffffffffffffff0
0x10e859348    a9bf79fd               stp fp, lr, [r15, #-16 ]!
0x10e85934c    aa0f03fd               mov fp, r15
        ;; v2 <- AllocateObject(Text) T{Text}
0x10e859350    94000000               bl 0x10e859350
        ;; ParallelMove r0 <- r0
        ;; StoreInstanceField(v2 . data = v4, barrier removed)
        ;; Inlined [Text.Text.]
0x10e859354    91403371               addi ip1, pp, 0xc000
0x10e859358    f944a631               ldrx ip1, [ip1, #2376]
0x10e85935c    f800f011               strx ip1, [r0, #15]
        ;; StoreInstanceField(v2 . style = v0, barrier removed)
        ;; Inlined [Text.Text.]
0x10e859360    f9405f51               ldrx ip1, [thr, #184]   null
0x10e859364    f801f011               strx ip1, [r0, #31]
        ;; StoreInstanceField(v2 . strutStyle = v0, barrier removed)
        ;; Inlined [Text.Text.]
0x10e859368    f9405f51               ldrx ip1, [thr, #184]   null
0x10e85936c    f8027011               strx ip1, [r0, #39]
        ;; StoreInstanceField(v2 . textAlign = v0, barrier removed)
        ;; Inlined [Text.Text.]
0x10e859370    f9405f51               ldrx ip1, [thr, #184]   null
0x10e859374    f802f011               strx ip1, [r0, #47]
        ;; StoreInstanceField(v2 . textDirection = v0, barrier removed)
        ;; Inlined [Text.Text.]
0x10e859378    f9405f51               ldrx ip1, [thr, #184]   null
0x10e85937c    f8037011               strx ip1, [r0, #55]
        ;; StoreInstanceField(v2 . locale = v0, barrier removed)
        ;; Inlined [Text.Text.]
0x10e859380    f9405f51               ldrx ip1, [thr, #184]   null
0x10e859384    f803f011               strx ip1, [r0, #63]
        ;; StoreInstanceField(v2 . softWrap = v0, barrier removed)
        ;; Inlined [Text.Text.]
0x10e859388    f9405f51               ldrx ip1, [thr, #184]   null
0x10e85938c    f8047011               strx ip1, [r0, #71]
        ;; StoreInstanceField(v2 . overflow = v0, barrier removed)
        ;; Inlined [Text.Text.]
0x10e859390    f9405f51               ldrx ip1, [thr, #184]   null
0x10e859394    f804f011               strx ip1, [r0, #79]
        ;; StoreInstanceField(v2 . textScaleFactor = v0, barrier removed)
        ;; Inlined [Text.Text.]
0x10e859398    f9405f51               ldrx ip1, [thr, #184]   null
0x10e85939c    f8057011               strx ip1, [r0, #87]
        ;; StoreInstanceField(v2 . maxLines = v0, barrier removed)
        ;; Inlined [Text.Text.]
0x10e8593a0    f9405f51               ldrx ip1, [thr, #184]   null
0x10e8593a4    f805f011               strx ip1, [r0, #95]
        ;; StoreInstanceField(v2 . semanticsLabel = v0, barrier removed)
        ;; Inlined [Text.Text.]
0x10e8593a8    f9405f51               ldrx ip1, [thr, #184]   null
0x10e8593ac    f8067011               strx ip1, [r0, #103]
        ;; StoreInstanceField(v2 . textWidthBasis = v0, barrier removed)
        ;; Inlined [Text.Text.]
0x10e8593b0    f9405f51               ldrx ip1, [thr, #184]   null
0x10e8593b4    f806f011               strx ip1, [r0, #111]
        ;; StoreInstanceField(v2 . key = v0, barrier removed)
        ;; Inlined [Text.Text. -> StatelessWidget.StatelessWidget. -> Widget.Widget.]
0x10e8593b8    f9405f51               ldrx ip1, [thr, #184]   null
0x10e8593bc    f8007011               strx ip1, [r0, #7]
        ;; ParallelMove r0 <- r0
        ;; Return:16(v2)
0x10e8593c0    aa1d03ef               mov r15, fp
0x10e8593c4    a8c179fd               ldp fp, lr, [r15], #16 !
0x10e8593c8    d65f03c0               ret lr
0x10e8593cc    d4200000               brk 0x0
```

out of 36 instructions 24 are actually redundant. Each store of `null` value is not needed because object is allocated `null` initialized. 

```
        ;; StoreInstanceField(v2 . key = v0, barrier removed)
0x10e8593b8    f9405f51               ldrx ip1, [thr, #184]   null
0x10e8593bc    f8007011               strx ip1, [r0, #7]
```

This has been known for a long time - but somehow we never got to fix it. I filed dart-lang/sdk#38454
 to track this work an immediate fix should be rather simple.

This would leave 12 instructions in the function which would actually be smaller than `func1` in Objective-C (16 instructions). 

```
  stp x29, x30, [sp, #-16]!   ; 8-byte Folded Spill
  mov x29, sp
  adrp  x8, l_OBJC_CLASSLIST_REFERENCES_$_@PAGE
  ldr x0, [x8, l_OBJC_CLASSLIST_REFERENCES_$_@PAGEOFF]
  adrp  x8, l_OBJC_SELECTOR_REFERENCES_@PAGE
  ldr x1, [x8, l_OBJC_SELECTOR_REFERENCES_@PAGEOFF]
  bl  _objc_msgSend
  adrp  x8, l_OBJC_SELECTOR_REFERENCES_.2@PAGE
  ldr x1, [x8, l_OBJC_SELECTOR_REFERENCES_.2@PAGEOFF]
  fmov  d0, xzr
  fmov  d1, xzr
  fmov  d2, xzr
  fmov  d3, xzr
  bl  _objc_msgSend
  ldp x29, x30, [sp], #16     ; 8-byte Folded Reload
  b _objc_autoreleaseReturnValue
```

Though Objective-C function does not actually do the same thing as Dart function does. 

@limengyun2008 could you post some Objective-C (or Swift) code that does something equivalent to what `func3` does in Dart code, i.e. Objective-C code that creates a column with few text widgets and an icon. I would really like to compare the size of that code to Dart code.

**Comment by limengyun2008** [comment]
@mraleph sure. but need some time cause I‘m outside now...will post as soon as possible. 

actually, in case of func1, the fact that dart(after removing redundant instructions)  beats oc is not surprising at all: the oc version is more complicated than the dart version. oc version has created two instances and CGRectMake takes 4 args，whereas the dart version only creates one instance.

so I guess the result will be the same as func1?  func3 just creates 6 instances(5 normal widget and 1 list)

after all, I will post a more rational oc version(both func1 and func3) tomorrow.

thanks for your reply, You convinced me this problem can really be solved.

**Comment by limengyun2008** [comment]
@mraleph 

here is the updated example:

```Objective-C
@interface CustomView1 : UILabel

- (instancetype)initWithString:(NSString *)text;

@end

@interface CustomView2 : UIImageView

- (instancetype)initWithImageName:(NSString *)imageName;

@end

@implementation CustomView1

- (instancetype)initWithString:(NSString *)text {
    if (self = [super init]) {
        self.text = text;
    }
    
    return self;
}

@end

@implementation CustomView2

- (instancetype)initWithImageName:(NSString *)imageName
{
    if (self = [super init]) {
        self.image = [UIImage imageNamed:imageName];
    }
    
    return self;
}

@end


@implementation ViewController

- (UIView *)func1 {
    return [[CustomView1 alloc] initWithString:@"blablabla"];
}

- (UIView *)func2 {
    return [[CustomView1 alloc] initWithString:@"blablabla"];
}

- (UIView *)func3 {
    return [[UIStackView alloc] initWithArrangedSubviews:@[[[CustomView1 alloc] initWithString:@"blablabla"],
                                                           [[CustomView1 alloc] initWithString:@"blablabla"],
                                                           [[CustomView1 alloc] initWithString:@"blablabla"],
                                                           [[CustomView2 alloc] initWithImageName:@"add"]]];
}



- (void)viewDidLoad {
    [super viewDidLoad];
    [self func1];
    [self func3];
    // Do any additional setup after loading the view, typically from a nib.
}

@end

```

Thanks!

**Comment by mraleph** [comment]
I have looked specifically at `func3` in Objective-C vs Dart. 

In Objective-C we get 99 instructions for the last version you provided. 

In Dart (after [cl/118286](https://dart-review.googlesource.com/c/sdk/+/118286) lands) we get 92 instructions. Out of those 92 instructions the biggest contributor to code size are write barriers on `StoreIndexed` instructions. 

So overall I think when these low hanging fruits are fixed (reducing alignment of entry points and removing instruction headers) - we seem to do pretty well compared to Objective-C in terms of raw code size. 

However we still probably emit more metadata associated with code than Objective-C, for example for `func3` we will emit 9 `StackMap` objects (some of which can be canonicalised with other code objects - some might not). 

```
StackMaps for function 'package:hello_world/main.dart_::_func3' {
0x018: 000000000000
0x01c: 00
0x048: 10
0x050: 11
0x08c: 11
0x0c8: 11
0x104: 11
0x148: 10
0x15c: 10
}
```

So I think the next big reduction will come from reducing metadata overheads. This is tracked by: dart-lang/sdk#35851 and dart-lang/sdk#35274

**Comment by limengyun2008** [comment]
excellent analysis!

seems we can expect a huge cut after all these five issues are resolved:

-   dart-lang/sdk#38451
-   dart-lang/sdk#38452
-   dart-lang/sdk#38454
-   dart-lang/sdk#35851
-   dart-lang/sdk#35274

can we expect all these work will be released in the next flutter stable release?

**Comment by mraleph** [comment]
Circling back to this issue. We have landed a number of fixes in the Dart AOT compiler. Now when I repeat the test adding 9 times `func1`, `func2` and `func3` - snapshot size grows by 7915 bytes. Similar experiment with Objective-C functions that you have provided yields 3260 bytes in growth. 

Still some way to go - but much closer now.

**Comment by mraleph** [comment]
I was measuring things again to check how far did we get, as part of this I created a somewhat more automated test based on @limengyun2008 code. This revealed a very interesting thing, which I have overlooked before because I was focusing too much on Dart AOT compiler output: 

**MACH-O format seems to have some alignment requirements which make raw binary size jump 16KB whenever your text segment passes certain limits. This happens irrespective of whether you use Dart or Objective-C - though with Dart it might happen at different moments (more often) because Dart produces a bit more code (~20-40%).**

Here is what I have measured (my scripts are all in [this repo](https://github.com/mraleph/objc-flutter-code-size), but I also have modified Flutter tools to pass `--dwarf-stack-traces` and `--obfuscate` to the `gen_snapshot` and strip resulting framework) following @limengyun2008 method of duplicating `func1`, `func2` and `func3`.

| # added copies  | ObjC Mach-O growth | ObjC Sections growth | Flutter Mach-O growth | Flutter Sections growth |
| ------------- | ------------- | ------------- | ------------- | ------------- |
|1|508|774|83312|93504|
|2|136|554|16592|784|
|3|136|554|0|672|
|4|120|554|0|736|
|5|136|554|0|800|
|6|128|554|16|736|
|7|136|554|0|688|
|8|120|554|0|768|
|9|136|554|0|720|
|10|128|554|0|736|
|11|136|558|16|752|
|12|128|558|0|736|
|13|136|558|0|688|
|14|128|558|0|736|
|15|144|558|0|736|
|16|128|558|0|736|
|17|136|558|16|768|
|18|128|558|0|720|
|19|136|558|0|640|
|20|128|558|0|736|
|21|136|558|0|736|
|22|128|558|16|768|
|23|144|558|0|704|
|24|128|558|16592|704|
|25|136|558|0|672|
|26|128|558|0|688|
|27|136|558|16|752|
|28|128|558|0|672|
|29|136|558|0|784|
|30|128|558|0|880|
|31|144|558|0|704|
|32|128|558|0|768|
|33|136|558|16|752|
|34|128|558|0|800|
|35|16520|558|0|752|
|36|128|558|0|752|
|37|136|558|0|752|
|38|128|558|16|704|
|39|144|558|0|720|
|40|128|558|0|704|
|41|136|558|0|720|
|42|128|558|0|704|
|43|136|558|16|768|
|44|128|558|0|752|
|45|136|558|0|704|
|46|128|558|16592|864|
|47|144|558|0|688|
|48|128|558|0|736|
|49|136|558|16|736|
|50|128|558|0|720|

As you can see even if you use Obj-C the file size would jump by 16KB every now and again. Interestingly enough with Obj-C application binary size grows continuously - while with Flutter it does not actually grow. 

Given the nature of 16KB jump I would say that it should not really matter - because the added space is most likely just padding bytes and those should compress very well. 

`--obfuscate` and `--dwarf-stack-traces` reduce overhead of adding a copy of `func*` by approximately ~18%. Here is the same table but with vanilla Flutter master:

| # added copies  | ObjC Mach-O growth | ObjC Sections growth | Flutter Mach-O growth | Flutter Sections growth |
| ------------- | ------------- | ------------- | ------------- | ------------- |
|1|508|774|99904|100880|
|2|136|554|0|896|
|3|136|554|0|880|
|4|120|554|0|880|
|5|136|554|16|928|
|6|128|554|16592|816|
|7|136|554|0|800|
|8|120|554|0|912|
|9|136|554|0|864|

This shows that in terms of pure bytes added to actual sections in the Mach-O file Dart you pay around 60% more right now. 

Given these numbers I think we continue focusing on making `--dwarf-stack-traces` (dart-lang/sdk#35851) and something that drops symbolic information from snapshots similar to obfuscation (dart-lang/sdk#38759) default in release builds.

However I also think this also reveals that original problem (to quote @limengyun2008 **the size do not increase linearly, it will suddenly grow 16KB**) is not actually a problem inherent to Flutter in any way - and Obj-C shows exactly the same pattern of growth.

**Comment by mraleph** [comment]
Mach-O documentation says the following:

> A Mach-O binary is organized into segments. Each segment contains one or more sections. Code or data of different types goes into each section. Segments always start on a page boundary, but sections are not necessarily page-aligned. The size of a segment is measured by the number of bytes in all the sections it contains and rounded up to the next virtual memory page boundary.

On ARM64 iOS page size is 16KB - which explains why binary size jumps in 16KB increments.

**Comment by Hixie** [comment]
@mraleph That updates is awesome, thanks! Any idea what time frame we should plan around here? (I'm looking for some guidance we can give our customers, so conservative targets are much preferred than targets that might slip, even if they end up being months more pessimistic than reality.)

**Comment by mraleph** [comment]
@Hixie All necessary machinery for `--dwarf-stack-traces` has landed, so the ball around that is now on the Flutter side to figure out how to roll that. I think the next thing that lands early in January would be removal of `RawInstruction` headers which would bring some (but not gigantic) reduction in instruction portion.  Reduction in the symbolic information would take some time - I would say we are looking at changes landing during the first two months of Q1.

**Comment by Hixie** [comment]
@mraleph is there a doc that talks about what flutter needs to do to support `--dwarf-stack-traces`?

**Comment by mkustermann** [comment]
> @mraleph is there a doc that talks about what flutter needs to do to support --dwarf-stack-traces?

Yes, @sstrickl is writing it up atm - we'll circulate it once it's ready.

**Comment by xster** [comment]
Also cc @dnfield for helping roll/package improvements into Flutter if needed.

**Comment by Hixie** [comment]
fantastic thanks

**Comment by dnfield** [comment]
@jonahwilliams FYI.  I'm assuming this will be a new argument either in the front end compiler or the tool.

**Comment by Hixie** [comment]
@sstrickl If my understanding is correct, we need to add support for https://github.com/dart-lang/sdk/blob/master/runtime/docs/dwarf_stack_traces.md to the Flutter command line tool and that will unblock some additional changes to save further binary size on iOS? Is that right?

**Comment by jonahwilliams** [comment]
@Hixie I filled https://github.com/flutter/flutter/issues/48726 and https://github.com/flutter/flutter/issues/48725 to enable AOT stripping and no-casual/lazy-async stack traces respectively. I can increase the priority of those issues to match here

**Comment by jonahwilliams** [comment]
This would also be blocked by the availability of the library to translate stack traces: https://github.com/dart-lang/sdk/blob/1eacac48cf/pkg/vm/bin/convert_stack_traces.dart

**Comment by Hixie** [comment]
Ok the status is that there's two bugs left that we're planning on working on, one is https://github.com/dart-lang/sdk/issues/38451 and the other is https://github.com/flutter/flutter/issues/48726. Should we close this bug?

**Comment by mraleph** [comment]
We will continue working on the code size and performance improvements (there are more planned things in the pipeline - the most comprehensive way to track this is to look at [vm-aot-code-size](https://github.com/dart-lang/sdk/issues?q=is%3Aopen+is%3Aissue+label%3Avm-aot-code-size) label in the Dart SDK repository).

I think we can close this issue especially because we now understand the reason why "the size do not increase linearly, it will suddenly grow 16KB" - which as it turns out has nothing to do with Dart.

**Comment by zoeyfan** [comment]
Sounds good to me.

**Comment by mkustermann** [comment]
Re-running @mraleph 's size comparison tool from [mraleph/objc-flutter-code-size](https://github.com/mraleph/objc-flutter-code-size) (with a small [fix](https://github.com/mraleph/objc-flutter-code-size/pull/1) to make it work for me) on flutter master (commit flutter/flutter@721927efd3ae583e76a55071e83de63ff1166e0b) results in these new numbers.

For brevity I've added `Flutter Default` and `Flutter SizeOpt` in the same table. The latter `Flutter SizeOpt` is explicitly opting into using size-optimizing flags (`--obfuscate,--dwarf-stack-traces,--strip`)

| # added copies  | ObjC Mach-O growth | ObjC Sections growth | Flutter Default Mach-O growth | Flutter Default Sections growth | Flutter SizeOpt Mach-O growth | Flutter SizeOpt Sections growth |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |                                      
|1|588|750|57112|57648|54112|46656|
|2|256|482|96|704|88|528| 
|3|168|470|16472|800|80|560|
|4|128|470|96|752|88|528| 
|5|136|470|88|832|80|560|
|6|160|466|88|784|80|544|
|7|136|466|96|784|88|528|
|8|120|466|88|800|80|560|
|9|136|466|96|736|88|544|
|10|128|466|88|752|80|528|
|11|136|470|88|784|80|544|
|12|128|470|96|768|88|544|
|13|136|470|96|912|80|528|
|14|128|470|88|768|80|560|
|15|144|470|96|736|88|528|
|16|128|470|88|688|80|560|
|17|136|470|104|736|88|528|
|18|128|470|88|752|80|592|
|19|136|470|88|720|80|544|
|20|128|470|96|688|88|560|
|21|136|470|96|768|80|544|
|22|128|470|88|768|88|544|
|23|144|470|96|768|16464|544|
|24|128|470|88|752|80|544|
|25|136|470|16488|800|88|544|
|26|128|470|88|736|80|528|
|27|136|470|88|768|80|576|
|28|128|470|96|336|88|528|
|29|136|470|96|752|80|544|
|30|128|470|88|912|88|512|
|31|144|470|96|752|80|544|
|32|128|470|88|784|80|528|
|33|136|470|104|784|88|544|
|34|128|470|88|752|80|528|
|35|136|470|88|768|80|544|
|36|128|470|96|784|88|544|
|37|136|470|96|752|80|544|
|38|128|470|96|784|88|544|
|39|144|470|88|784|80|528|
|40|128|470|88|768|80|528|
|41|136|470|104|800|88|512|
|42|128|470|88|736|16464|8736|
|43|16520|470|96|800|80|560|
|44|128|470|88|768|88|544|
|45|136|470|96|720|80|528|
|46|128|470|96|800|88|544|
|47|144|470|16472|768|80|544|
|48|128|470|88|736|80|544|
|49|136|470|104|816|88|544|
|50|128|470|88|752|80|528|


If we compare the `Flutter SizeOpt` numbers from the first table in https://github.com/flutter/flutter/issues/40345#issuecomment-563304914 (december 9th) we can see a big drop from:

| # added copies  | ObjC Mach-O growth | ObjC Sections growth | Flutter Default Mach-O growth | Flutter Default Sections growth | Flutter SizeOpt Mach-O growth | Flutter SizeOpt Sections growth |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
1 | 508 | 774 | 99904 | 100880 | 83312 | 93504
2 | 136 | 554 | 16592 | 784
3 | 136 | 554 | 0 | 896 | 0 | 672
4 | 120 | 554  | 0 | 880 | 0 | 736
5 | 136 | 554 | 16 | 928 | 0 | 800
...
45 | 136 | 558 | - | - | 0 | 704
46 | 128 | 558 | - | -  | 16592 | 864
47 | 144 | 558  | - | - | 0 | 688
48 | 128 | 558  | - | - | 0 | 736
49 | 136 | 558  | - | - | 16 | 736
50 | 128 | 558  | - | - | 0 | 720

to

| # added copies  | ObjC Mach-O growth | ObjC Sections growth | Flutter Default Mach-O growth | Flutter Default Sections growth | Flutter SizeOpt Mach-O growth | Flutter SizeOpt Sections growth |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
|1|588|750|57112|57648|54112|46656|
|2|256|482|96|704|88|528| 
|3|168|470|16472|800|80|560|
|4|128|470|96|752|88|528| 
|5|136|470|88|832|80|560|
...
|45|136|470|96|720|80|528|
|46|128|470|96|800|88|544|
|47|144|470|16472|768|80|544|
|48|128|470|88|736|80|544|
|49|136|470|104|816|88|544|
|50|128|470|88|752|80|528|

=> Focusing on the `ObjC Sections growth` column, we can see a drop of roughly 15%. This is due to improvements Objective-C compiler (seemingly mainly due to added support of outlining in LLVM, which Dart doesn't do at the moment)
=> Focusing on the `Flutter SizeOpt Sections growth` column, we can see a drop of roughly 25%. This is due to improvements to the Flutter/Dart AOT compiler.

Comparing the two we see `Flutter SizeOpt` being  maybe around 14% larger than ObjC and `Flutter Default` being maybe around 65% larger than ObjC.

=> In summary: When turning on size optimizing flags we are quite close to Objective-C (in this particular size benchmark).

**Comment by mkustermann** [comment]
Re-running the size comparison tool (with a few small fixes) on flutter master (commit 1dc458b2fe) results in these new numbers.

| # added copies  | ObjC Mach-O growth | ObjC Sections growth | Flutter Default Mach-O growth | Flutter Default Sections growth | Flutter SizeOpt Mach-O growth | Flutter SizeOpt Sections growth |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
|1|584|750|21760|30016|36592|23472|
|2|256|482|96|720|80|384|
|3|168|470|16672|720|96|336|
|4|128|470|96|656|80|320|
|5|136|470|96|784|80|336|
|...|||||||
|45|144|470|96|672|96|352|
|46|128|470|80|656|80|304|
|47|136|470|112|656|80|336|
|48|128|470|80|672|80|320|
|49|136|470|96|624|80|320|
|50|128|470|96|688|80|320|

=> This seems to indicate that on this particular benchmark Flutter is smaller than the corresponding ObjC code in size-optimized mode!

**Comment by xster** [comment]
Woohoo, that's amazing!
